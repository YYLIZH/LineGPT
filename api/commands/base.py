import typing
from abc import ABC

from api.utils.configs import LANGUAGE


class Command(ABC):
    usage_en = ""
    usage_zh = ""

    def __init__(
        self, subcommand: typing.Optional[str] = None, args: typing.Optional[str] = None
    ) -> None:
        self.args = args

    def execute(self, **kwargs):
        raise NotImplementedError

    def print_usage(cls):
        return getattr(cls, f"usage_{LANGUAGE}")
