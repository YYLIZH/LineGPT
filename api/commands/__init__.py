from collections import namedtuple
from typing import Dict
from api.utils.configs import LANGUAGE

CommandInfo = namedtuple(
    "CommandInfo",
    "module_path, class_name, summary_en, summary_zh_TW",
)
commands_info: Dict[str, CommandInfo] = {
    "gpt": CommandInfo(
        "api.commands.gpt", "GptCommand", "GPT commands", "GPT 指令"
    ),
    "weather": CommandInfo(
        "api.commands.weather",
        "WeatherCommand",
        "Check the weather",
        "查看天氣",
    ),
    "settle": CommandInfo(
        "api.commands.settle",
        "SettleCommand",
        "Settle the expenses among a group",
        "多人分帳",
    ),
    "eat": CommandInfo(
        "api.commands.eat",
        "EatCommand",
        "What to eat now",
        "現在吃什麼",
    ),
}


def print_usage():
    heading = (
        "嗨！我是LineGPT，很高興為您服務。以下是我的使用說明：\n\n"
        "@LineGPT <指令>\n"
        "可用指令\n"
    )
    footer = "\n您可以使用'@LineGPT <指令> help'來查看更多資訊"
    if LANGUAGE == "en":
        heading = (
            "Hi! I am LineGPT. It's my pleasure to help you. Here is the usage:\n\n"
            "@LineGPT <Command>\n"
            "Available commands\n"
        )
        footer = "\nYou can use '@LineGPT <Command> help' to get more information about each command."

    res = heading
    for key, value in commands_info.items():
        summary = getattr(value, f"summary_{LANGUAGE}")
        line = f"{key:<15}{summary}\n"
        res += line
    res += footer
    return res
