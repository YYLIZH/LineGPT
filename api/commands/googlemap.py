from datetime import datetime, timedelta
from typing import Literal


class GoogleMapSession:
    timeout = 180

    def __init__(self):
        self.last_update_time = datetime.now() - timedelta(days=1)
        self.app = None

    def is_expired(self) -> bool:
        expired_time = self.last_update_time + timedelta(
            seconds=self.timeout
        )
        if datetime.now() > expired_time:
            return True
        return False

    def reset(self):
        self.app = None
        self.last_update_time = datetime.now() - timedelta(days=1)

    def set_app(self, app: Literal["eat", "toilet"]):
        self.last_update_time = datetime.now()
        self.app = app


GOOGLE_MAP_SESSION = GoogleMapSession()
