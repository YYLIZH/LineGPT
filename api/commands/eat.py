from __future__ import annotations

import math
import textwrap
from datetime import datetime, timedelta
from enum import Enum

import requests

from api.commands.base import Command
from api.utils.configs import GOOGLE_API_KEY, LANGUAGE


class MessageEN(str, Enum):
    REPLY = "Please share your location to me"
    NO_RESULT = "No result in nearby location."


class MessageZHTW(str, Enum):
    REPLY = "請分享你的位置資訊給我"
    NO_RESULT = "找不到結果"


MESSAGE = MessageEN if LANGUAGE == "en" else MessageZHTW


def get_food(latitude: str, longitude: str) -> list[dict]:
    radius = 3000
    place_type = "restaurant"
    location = f"{latitude},{longitude}"

    query_opts = [
        f"location={location}",
        f"radius={radius}",
        f"type={place_type}",
        f"key={GOOGLE_API_KEY}",
        "opennow=true",
    ]
    if LANGUAGE == "zh_TW":
        query_opts.append("language=zh-TW")

    url = f"https://maps.googleapis.com/maps/api/place/nearbysearch/json?" + "&".join(
        query_opts
    )

    try:
        response = requests.get(url)
        data = response.json()
        return sorted(data["results"], key=lambda x: x["rating"], reverse=True)[:10]
    except Exception:
        return []


def create_map_link(latitude: str, longitude: str, place_id: str) -> str:
    map_link = f"https://www.google.com/maps/search/?api=1&query={latitude}%2C{longitude}&query_place_id={place_id}"
    return map_link

def calculate_distance(lat1:float, lon1:float, lat2:float, lon2:float)->str:
    """Use Haversine Formula"""
    # 地球半徑（公里）
    R = 6371.0
    
    # 將經緯度轉換為弧度
    lat1_rad = math.radians(lat1)
    lon1_rad = math.radians(lon1)
    lat2_rad = math.radians(lat2)
    lon2_rad = math.radians(lon2)
    
    # 經度和緯度的差值
    dlat = lat2_rad - lat1_rad
    dlon = lon2_rad - lon1_rad
    
    # 哈弗辛公式
    a = math.sin(dlat / 2)**2 + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    # 距離（公里）
    distance = R * c

    if distance>1:
        return f"{round(distance,2)} km"

    return f"{round(distance*1000,2)} m"

def get_place_detail(user_lat:str,user_lng:str,place_info: dict) -> dict[str, str]:
    name = str(place_info.get("name", ""))
    rating = place_info.get("rating", 0.0)
    user_ratings_total = place_info.get("user_ratings_total", 0)
    address = str(place_info.get("vicinity", "N/A"))

    latitude = place_info["geometry"]["location"]["lat"]
    longitude = place_info["geometry"]["location"]["lng"]

    return {
        "name": name,
        "address": address,
        "distance":calculate_distance(float(user_lat),float(user_lng),latitude,longitude),
        "rating": rating,
        "user_ratings_total": user_ratings_total,
        "map_url": create_map_link(latitude, longitude, place_info["place_id"]),
    }


def generate_card(
    name: str, address:str,distance:float,rating: float, user_ratings_total: int, map_url: str
) -> dict:
    card = {
        "type": "bubble",
        "size": "micro",
        "body": {
            "type": "box",
            "layout": "vertical",
            "contents": [
                {
                    "type": "text",
                    "text": str(name),
                    "weight": "bold",
                    "size": "sm",
                    "wrap": True,
                },
                {
                    "type": "box",
                    "layout": "baseline",
                    "contents": [],
                },
                {
                    "type": "box",
                    "layout": "baseline",
                    "contents": [
                        {
                            "type": "text",
                            "text": f"{user_ratings_total} votes",
                            "size": "sm",
                        }
                    ],
                },
                {
                    "type": "box",
                    "layout": "vertical",
                    "contents": [
                        {
                            "type": "box",
                            "layout": "baseline",
                            "spacing": "sm",
                            "contents": [
                                {
                                    "type": "text",
                                    "text": str(address),
                                    "wrap": True,
                                    "color": "#8c8c8c",
                                    "size": "xs",
                                    "flex": 5,
                                }
                            ],
                        },
                        {
                            "type": "box",
                            "layout": "baseline",
                            "spacing": "sm",
                            "contents": [
                                {
                                    "type": "text",
                                    "text": str(distance),
                                    "wrap": True,
                                    "color": "#8c8c8c",
                                    "size": "xs",
                                    "flex": 5,
                                }
                            ],
                        }
                    ],
                },
            ],
            "spacing": "sm",
            "paddingAll": "13px",
        },
        "footer": {
            "type": "box",
            "layout": "horizontal",
            "contents": [
                {
                    "type": "button",
                    "action": {
                        "type": "uri",
                        "label": "Go",
                        "uri": map_url,
                    },
                    "style": "primary",
                }
            ],
        },
    }

    # calculate rating box content
    rating_box_content = []
    rating_floor = math.ceil(rating)
    num_golden = rating_floor
    num_empty = 5 - num_golden
    for _ in range(num_golden):
        rating_box_content.append(
            {
                "type": "icon",
                "size": "xs",
                "url": "https://developers-resource.landpress.line.me/fx/img/review_gold_star_28.png",
            }
        )
    for _ in range(num_empty):
        rating_box_content.append(
            {
                "type": "icon",
                "size": "xs",
                "url": "https://developers-resource.landpress.line.me/fx/img/review_gray_star_28.png",
            }
        )

    rating_box_content.append(
        {
            "type": "text",
            "text": str(rating),
            "size": "xs",
            "color": "#8c8c8c",
            "margin": "md",
            "flex": 0,
        }
    )
    card["body"]["contents"][1]["contents"] = rating_box_content
    return card


def generate_flex_message(place_details: list[dict]) -> dict:
    flex_message_template = {
        "type": "carousel",
        "contents": [],
    }

    for detail in place_details:
        flex_message_template["contents"].append(generate_card(**detail))

    return flex_message_template


def what_to_eat(latitude: str, longitude: str) -> dict:
    restaurants = get_food(latitude, longitude)
    if len(restaurants) == 0:
        return MESSAGE.NO_RESULT.value

    # Format the text
    place_details = []
    for restaurant in restaurants:
        place_detail = get_place_detail(latitude,longitude,restaurant)
        place_details.append(place_detail)

    flex_message_dict = generate_flex_message(place_details)

    return flex_message_dict


class GoogleMapSession:
    timeout = 5

    def __init__(self):
        self.last_update_time = datetime.now() - timedelta(days=1)

    def update_time(self):
        self.last_update_time = datetime.now()

    def is_expired(self) -> bool:
        expired_time = self.last_update_time + timedelta(minutes=self.timeout)
        if datetime.now() > expired_time:
            return True
        return False


class EatCommand(Command):
    usage_en = textwrap.dedent(
        """
        * Find opening restaurant nearby.

        Example:
        @LineGPT eat
        """
    )

    usage_zh_TW = textwrap.dedent(
        """
        * 尋找附近的餐廳

        Example: 
        @LineGPT eat
        """
    )

    def __init__(self, subcommand=None, args=None):
        super().__init__(subcommand, args)
        self.google_map_session = None

    @classmethod
    def setup(cls, args_msg: str) -> EatCommand:
        return cls(None, args_msg)

    def load(self, google_map_session: GoogleMapSession):
        self.google_map_session = google_map_session

    def execute(self, **kwargs):
        return MESSAGE.REPLY.value
