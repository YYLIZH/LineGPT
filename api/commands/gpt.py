from __future__ import annotations
import datetime
import re
import typing
from enum import Enum
from typing import Dict

import openai

from api.commands.base import Command
from api.utils.configs import (
    LANGUAGE,
    OPENAI_API_KEY,
    OPENAI_FREQUENCY_PENALTY,
    OPENAI_MAX_TOKEN,
    OPENAI_MODEL,
    OPENAI_PRESENCE_PENALTY,
    OPENAI_TEMPERATURE,
    SESSION_EXPIRED,
)
from api.utils.info import Error, Warning


class MessageEN(Enum):
    GREETING = "Start the dialogue session."
    FOUND_SESSION = "Found existing session. You can see existing session's dialog, keep talking, or restart a new session."
    NO_SESSION = "Session does not exist. Please start the session first."
    SESSION_EXPIRED = (
        "Current dialogue session is expired. Please restart the dialogue session."
    )
    RUNTIME_ERROR = "RuntimeError. Please ask again later."
    UNEXPECTED_ERROR = "Unexpected error happened. Restart the session."
    CLOSE = "Close current dialogue session."
    RESTART = "Restart the dialogue session"
    LOG = "Below are the past dialogues:"
    NOT_ALLOW_METHOD = "This method is not allowed."


class MessageZHTW(Enum):
    GREETING = "對話階段已開始"
    FOUND_SESSION = "存在舊有的對話階段，您可以查看紀錄、繼續對話、或是重啟對話階段。"
    NO_SESSION = '尚未開啟對話階段，請先使用"@LineGPT start"開啟對話。'
    SESSION_EXPIRED = '當前對話階段已過期。請先使用"@LineGPT restart"重新開啟對話階段。'
    RUNTIME_ERROR = "RuntimeError. 請稍後再試一次"
    UNEXPECTED_ERROR = "發生未知的錯誤，重啟對話。"
    CLOSE = "關閉對話階段。"
    RESTART = "重新啟動對話階段"
    LOG = "以下是對話紀錄："
    NOT_ALLOW_METHOD = "不存在的指令"


USAGE_EN = """* Start a dialogue session.
@LineGPT gpt start

* Show dialogue
@LineGPT gpt log

* Ask a question
@LineGPT gpt talk <your question> 
Example:
@LineGPT gpt talk What is graphene?

* Close an existing dialogue session
@LineGPT gpt close
"""

USAGE_ZH_TW = """* 開始對話階段
@LineGPT gpt start

* 顯示過往對話紀錄
@LineGPT gpt log

* 提問
@LineGPT gpt talk <your question> 
Example:
@LineGPT gpt talk 請告訴我怎麼變有錢人

* 關閉對話階段，過往紀錄將會被刪除
@LineGPT gpt close
"""

openai.api_key = OPENAI_API_KEY
MESSAGE = MessageEN if LANGUAGE == "en" else MessageZHTW


class DialogueSession:
    def __init__(self) -> None:
        self.dialogue = []
        self.last_update_time = datetime.datetime.now()

    def add_ai_text(self, text: str) -> None:
        self.dialogue.append(f"AI: {text}")
        self.last_update_time = datetime.datetime.now()

    def add_human_text(self, text: str) -> None:
        self.dialogue.append(f"Human: {text}")
        self.last_update_time = datetime.datetime.now()

    def dump_dialogue(self) -> str:
        return "\n".join(self.dialogue)

    def is_expired(self) -> bool:
        time_delta: datetime.timedelta = datetime.datetime.now() - self.last_update_time
        if time_delta.seconds >= int(SESSION_EXPIRED):
            return True
        return False


class GPT:
    def __init__(self) -> None:
        self.dialogue_session = DialogueSession()
        self.model = OPENAI_MODEL
        self.temperature = OPENAI_TEMPERATURE
        self.max_tokens = OPENAI_MAX_TOKEN
        self.frequency_penalty = OPENAI_FREQUENCY_PENALTY
        self.presence_penalty = OPENAI_PRESENCE_PENALTY

    def _talk(self):
        response = openai.Completion.create(
            model=self.model,
            prompt=self.dialogue_session.dump_dialogue(),
            temperature=self.temperature,
            max_tokens=self.max_tokens,
            frequency_penalty=self.frequency_penalty,
            presence_penalty=self.presence_penalty,
        )
        text = response["choices"][0]["text"].strip()
        extracted_text = re.search(r"(\w+: +)?([\s\S]*)", text).group(2)
        extracted_text = extracted_text.replace("AI: ", "")
        self.dialogue_session.add_ai_text(extracted_text)
        return extracted_text

    def talk(self, text):
        if self.dialogue_session is None:
            return Warning(MESSAGE.NO_SESSION.value)
        if self.dialogue_session.is_expired():
            return Warning(MESSAGE.SESSION_EXPIRED.value)

        self.dialogue_session.add_human_text(text)
        try:
            response = self._talk()
        except Exception:
            self.dialogue_session.dialogue.pop()
            return Error(MESSAGE.RUNTIME_ERROR.value)
        return response


class GPTSessions:
    def __init__(self) -> None:
        """Each Group use different gpt."""
        self.sessions: Dict[str, GPT] = {}

    def start(self, group_id: str) -> str:
        """Start a GPT and add into session."""
        if gpt := self.sessions.get(group_id):
            return MESSAGE.FOUND_SESSION.value
        gpt = GPT()
        self.sessions[group_id] = gpt
        return MESSAGE.GREETING.value

    def close(self, group_id: str) -> str:
        self.sessions.pop(group_id)
        return MESSAGE.CLOSE.value

    def restart(self, group_id: str) -> str:
        gpt = GPT()
        self.sessions[group_id] = gpt
        return MESSAGE.GREETING.value

    def talk(self, group_id: str, text: str) -> str:
        if not self.sessions.get(group_id):
            return Warning(MESSAGE.NO_SESSION.value)
        gpt = self.sessions.get(group_id)
        try:
            return gpt.talk(text)
        except Exception:
            self.sessions[group_id] = GPT()
            return Error(MESSAGE.UNEXPECTED_ERROR.value)

    def log(self, group_id: str) -> str:
        gpt = self.sessions.get(group_id)
        return MESSAGE.LOG.value + "\n" + gpt.dialogue_session.dump_dialogue()


class GptCommand(Command):
    usage_en = USAGE_EN
    usage_zh_TW = USAGE_ZH_TW

    def __init__(
        self, subcommand: typing.Optional[str] = None, args: typing.Optional[str] = None
    ) -> None:
        super().__init__(subcommand, args)
        self.gpt_sessions: GPTSessions = None

    def load(self, gpt_sessions: GPTSessions):
        self.gpt_sessions = gpt_sessions

    @classmethod
    def setup(cls, args_msg: str) -> GptCommand:
        mrx = re.search(r"(\w+) *(.*)?", args_msg)
        subcommand, args = mrx.groups()
        return cls(subcommand, args)

    def execute(self, **kwargs):
        id = kwargs["id"]
        try:
            func = getattr(self.gpt_sessions, self.subcommand)
            if self.subcommand == "talk":
                return func(id, self.args)
            return func(id)
        except AttributeError:
            return Error(MESSAGE.NOT_ALLOW_METHOD.value)
