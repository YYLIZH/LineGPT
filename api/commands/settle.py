from __future__ import annotations
import re
import typing

from api.commands.base import Command
from api.utils.configs import LANGUAGE


class SettleCommand(Command):
    usage_en = """* Settle the money transaction
@LineGPT settle
<chart>

Example:
@LineGPT settle
Iron man: 100
Batman: 300
Superman: 240
Spiderman: 0

Here is the result:
Spiderman -> Batman 140.0
Iron man -> Superman 60.0
Spiderman -> Superman 20.0
"""

    usage_zh_TW = """* 多人分帳
@LineGPT settle
<chart>

Example:
@LineGPT settle
大壯: 100
小帥: 300
小美: 240
大黑: 0

以下是分帳結果：
大黑 -> 小帥 140.0
大壯 -> 小美 60.0
大黑 -> 小美 20.0
"""

    def __init__(
        self, subcommand: typing.Optional[str] = None, args: typing.Optional[str] = None
    ) -> None:
        args = args.replace("：", ":")  # Replace chinese full colon
        super().__init__(subcommand, args)
        self.expenses = {}

    @classmethod
    def setup(cls, args_msg: str) -> SettleCommand:
        args = args_msg.lstrip()
        return cls(None, args)

    def execute(self, **kwargs):
        self.parse_expense()
        transactions = self.settle_money()
        heading = "Here is the result:\n" if LANGUAGE == "en" else "以下是分帳結果：\n"

        return heading + "\n".join(
            [
                f"{transaction[0]} -> {transaction[1]} {transaction[2]}"
                for transaction in transactions
            ]
        )

    def parse_expense(self):
        pattern = re.compile(r"(.+): *(\d+)")
        for line in self.args.split("\n"):
            if mrx := pattern.search(line):
                self.expenses[mrx.group(1)] = int(mrx.group(2))

    def settle_money(self):
        """
        Calculates the amount each person owes or is owed after settling expenses.

        Arguments:
        expenses -- a dictionary of the form {person: amount_paid} representing the expenses paid by each person

        Returns:
        A list of tuples representing the money transactions needed to settle the expenses.
        """
        total_expenses = sum(self.expenses.values())
        num_people = len(self.expenses)
        average_expense = total_expenses / num_people
        owed = {
            person: amount_paid - average_expense
            for person, amount_paid in self.expenses.items()
        }

        transactions = []
        while any(owed.values()):
            person1, amount1 = max(owed.items(), key=lambda x: x[1])
            person2, amount2 = min(owed.items(), key=lambda x: x[1])
            amount = min(-amount2, amount1)
            transactions.append((person2, person1, amount))
            owed[person1] -= amount
            owed[person2] += amount

        return transactions
