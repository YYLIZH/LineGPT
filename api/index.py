import re
from importlib import import_module
from typing import Union

from fastapi import FastAPI, HTTPException, Request
from linebot import LineBotApi, WebhookHandler
from linebot.exceptions import InvalidSignatureError
from linebot.models import (
    FlexSendMessage,
    LocationMessage,
    MessageEvent,
    TextMessage,
    TextSendMessage,
)

from api.commands import commands_info, print_usage
from api.commands.base import Command
from api.commands.eat import EatCommand, GoogleMapSession, what_to_eat
from api.commands.gpt import GptCommand, GPTSessions
from api.utils.configs import (
    LINE_CHANNEL_ACCESS_TOKEN,
    LINE_CHANNEL_SECRET,
)
from api.utils.info import Error

app = FastAPI()
GPT_Sessions = GPTSessions()
GoogleMap_Session = GoogleMapSession()
line_bot_api = LineBotApi(LINE_CHANNEL_ACCESS_TOKEN)
line_handler = WebhookHandler(LINE_CHANNEL_SECRET)


def parse_message(message: str) -> Union[str, Command]:
    if re.match(r"^@LineGPT( help.*)?$", message):
        return print_usage()
    mrx = re.search(r"^@LineGPT +(\w+) *([\s\S]*)", message)
    if not mrx:
        return str(Error("Wrong format")) + "\n" + print_usage()
    command_str, other = mrx.groups()
    command_info = commands_info.get(command_str)
    if not command_info:
        return str(Error("No such command")) + "\n" + print_usage()
    module = import_module(command_info.module_path)
    command_cls: Command = getattr(module, command_info.class_name)
    command = command_cls.setup(other)
    if "help" in message:
        return command_cls.print_usage()
    if isinstance(command, GptCommand):
        command.load(GPT_Sessions)
    elif isinstance(command, EatCommand):
        GoogleMap_Session.update_time()
        command.load(GoogleMap_Session)
    return command


@app.get("/")
async def home(request: Request):
    return {"LineGPT": "Test"}


@app.get("/webhook")
async def checkWebhook(request: Request):
    return {"webhook": "OK"}


@app.post("/webhook")
async def LineGPTBot(request: Request):
    signature = request.headers["X-Line-Signature"]
    body = await request.body()
    try:
        line_handler.handle(body.decode(), signature)
    except InvalidSignatureError:
        raise HTTPException(status_code=400, detail="Missing Parameters")
    return "OK"


@line_handler.add(MessageEvent, message=TextMessage)
def handling_text_message(event):
    replyToken = event.reply_token
    if event.message:
        message: str = event.message.text
        if message.startswith("@LineGPT"):
            try:
                id = getattr(event.source, "group_id")
            except AttributeError:
                id = getattr(event.source, "user_id")
            parse_result = parse_message(message)
            if isinstance(parse_result, Command):
                command: Command = parse_result
                result = command.execute(**{"id": id})
            else:
                result = parse_result

            if result:
                echoMessages = TextSendMessage(text=str(result))
                line_bot_api.reply_message(
                    reply_token=replyToken, messages=echoMessages
                )


@line_handler.add(MessageEvent, message=LocationMessage)
def handling_location_message(event):
    replyToken = event.reply_token
    if event.message and GoogleMap_Session.is_expired() is False:
        latitude = event.message.latitude
        longitude = event.message.longitude
        result = what_to_eat(latitude, longitude)
        flex_message = FlexSendMessage("cards", result)
        line_bot_api.reply_message(reply_token=replyToken, messages=flex_message)
