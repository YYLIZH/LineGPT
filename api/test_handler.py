import pytest
from linebot.v3.messaging import LocationMessage, TextMessage

from api.commands import eat, print_usage, toilet

from . import line_handler


@pytest.fixture(name="fake_location")
def fixture_location_message():
    return LocationMessage(
        title="fake location",
        address="fake address",
        latitude=25.033611,
        longitude=121.564444,
    )


def test_print_usage(snapshot):
    help_message = print_usage()
    assert snapshot == help_message


def test_reply_text_message():
    text_message = TextMessage(text="aaa")
    messages = line_handler.reply_text_message(text_message)
    assert messages is None

    text_message = TextMessage(text="@LineGPT gogo")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == print_usage()

    text_message = TextMessage(text="@LineGPT help")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == print_usage()


def test_reply_text_message_eat(snapshot):
    text_message = TextMessage(text="@LineGPT eat")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == snapshot


def test_reply_text_message_gpt(snapshot):
    text_message = TextMessage(text="@LineGPT gpt")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == snapshot


def test_reply_text_message_settle(snapshot):
    text_message = TextMessage(text="@LineGPT settle")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == snapshot


def test_reply_text_message_weather(snapshot):
    text_message = TextMessage(text="@LineGPT weather")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == snapshot


def test_reply_text_message_toilet(snapshot):
    text_message = TextMessage(text="@LineGPT toilet")
    messages = line_handler.reply_text_message(text_message)
    assert messages[0].text == snapshot


def test_reply_location_message_no_existing_session(
    fake_location: LocationMessage,
):
    assert eat.GOOGLE_MAP_SESSION.is_expired()
    assert toilet.GOOGLE_MAP_SESSION.is_expired()
    messages = line_handler.reply_location_message(
        location_message=fake_location
    )
    assert messages is None


def test_reply_location_message_multiple_existing_session(
    snapshot, fake_location: LocationMessage
):
    eat.GOOGLE_MAP_SESSION.update_time()
    toilet.GOOGLE_MAP_SESSION.update_time()
    messages = line_handler.reply_location_message(
        location_message=fake_location
    )
    assert messages[0].text == snapshot
    eat.GOOGLE_MAP_SESSION.set_expired()
    toilet.GOOGLE_MAP_SESSION.set_expired()


def test_reply_location_message_eat(fake_location: LocationMessage):
    eat.GOOGLE_MAP_SESSION.update_time()
    messages = line_handler.reply_location_message(
        location_message=fake_location
    )
    print(messages)
    assert messages[0].alt_text == "restaurant cards"
    eat.GOOGLE_MAP_SESSION.set_expired()


def test_reply_location_message_toilet(
    fake_location: LocationMessage,
):
    toilet.GOOGLE_MAP_SESSION.update_time()
    messages = line_handler.reply_location_message(
        location_message=fake_location
    )
    assert messages[0].alt_text == "toilet cards"
    toilet.GOOGLE_MAP_SESSION.set_expired()
