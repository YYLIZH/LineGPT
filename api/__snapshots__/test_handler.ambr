# serializer version: 1
# name: test_print_usage
  '''
  嗨！我是LineGPT，很高興為您服務。以下是我的使用說明：
  
  @LineGPT <指令>
  可用指令
  gpt            GPT 指令
  weather        查看天氣
  settle         多人分帳
  eat            現在吃什麼
  
  您可以使用'@LineGPT <指令> help'來查看更多資訊
  '''
# ---
# name: test_reply_location_message_multiple_existing_session
  'The sessions of both eat and toilet are existing. Please stop one of them.'
# ---
# name: test_reply_text_message_eat
  '''
  
  * 開啟對話來尋找附近的餐廳
  
  Example: 
  @LineGPT eat start
  
  * 關閉目前對話來使用其他定位功能
  
  Example:
  @LineGPT eat stop
  
  '''
# ---
# name: test_reply_text_message_gpt
  '''
  
  * 開始對話階段
  @LineGPT gpt start
  
  * 顯示過往對話紀錄
  @LineGPT gpt log
  
  * 提問
  @LineGPT gpt talk <your question> 
  
  Example:
  @LineGPT gpt talk 請告訴我怎麼變有錢人
  
  * 關閉對話階段，過往紀錄將會被刪除
  @LineGPT gpt close
  
  * 重新開始對話階段
  @LineGPT gpt restart
  
  '''
# ---
# name: test_reply_text_message_settle
  '''
  
  * 多人分帳
  @LineGPT settle
  <chart>
  
  Example:
  @LineGPT settle
  大壯: 100
  小帥: 300
  小美: 240
  大黑: 0
  
  以下是分帳結果：
          大黑 -> 小帥 140.0
          大壯 -> 小美 60.0
          大黑 -> 小美 20.0
  
  '''
# ---
# name: test_reply_text_message_toilet
  '''
  
  * 開啟對話來尋找附近的廁所
  
  Example: 
  @LineGPT toilet start
  
  * 關閉目前對話來使用其他定位功能
  
  Example:
  @LineGPT toilet stop
  
  '''
# ---
# name: test_reply_text_message_weather
  '''
  
  * 查詢未來36小時的天氣預報
  @LineGPT weather <地點>
  
  Example:
  @LineGPT weather 嘉義縣
  
  '''
# ---
